perfkitbenchmarker/virtual_machine.py

```diff
@@ -181,99 +128,101 @@ def CreateScratchDisk(self, disk_spec):
-    else:
-      data_disk = disks[0]
+    if disk_num >= len(self.scratch_disks):
+      raise errors.Error(
+          'GetScratchDir(disk_num=%s) is invalid, max disk_num is %s' % (
+              disk_num, len(self.scratch_disks)))
+    return self.scratch_disks[disk_num].mount_point
 
-    self.scratch_disks.append(data_disk)
+  def GetLocalDisks(self):
+    # TODO(ehankland) This method should be removed as soon as raw/unmounted
+    # scratch disks are supported in a different way. Only the Aerospike
+    # benchmark currently accesses disks using this method.
+    """Returns a list of local disks on the VM."""
+    return []
 
-    if data_disk.disk_type != disk.LOCAL:
-      data_disk.Create()
-      data_disk.Attach(self)
+  def AddMetadata(self, **kwargs):
+    """Add key/value metadata to the instance.
 
-    if data_disk.is_striped:
-      device_paths = [d.GetDevicePath() for d in data_disk.disks]
-      self.StripeDisks(device_paths, data_disk.GetDevicePath())
+    Adds metadata in the form of key value pairs to the instance. Useful for
+    debugging / introspection.
 
-    if disk_spec.mount_point:
-      self.FormatDisk(data_disk.GetDevicePath())
-      self.MountDisk(data_disk.GetDevicePath(), disk_spec.mount_point)
+    The default implementation is a noop. Cloud providers supporting instance
+    metadata should override.
 
+    Args:
+      **kwargs: dict. (tag name, tag value) pairs to set as metadata on the
+        instance.
+    """
+    pass
 
-  def DeleteScratchDisks(self):
-    """Delete a VM's scratch disks."""
-    for scratch_disk in self.scratch_disks:
-      scratch_disk.Delete()
 
-  @vm_util.Retry(log_errors=False, poll_interval=1)
-  def WaitForBootCompletion(self):
-    """Waits until VM is has booted."""
-    resp, _ = self.RemoteCommand('hostname', retries=1, suppress_warning=True)
-    if self.bootable_time is None:
-      self.bootable_time = time.time()
-    if self.hostname is None:
-      self.hostname = resp[:-1]
-
-  @vm_util.Retry()
-  def FormatDisk(self, device_path):
-    """Formats a disk attached to the VM."""
-    fmt_cmd = ('sudo mke2fs -F -E lazy_itable_init=0 -O '
-               '^has_journal -t ext4 -b 4096 %s' % device_path)
-    self.RemoteCommand(fmt_cmd)
-
-  def MountDisk(self, device_path, mount_path):
-    """Mounts a formatted disk in the VM."""
-    mnt_cmd = ('sudo mkdir -p {1};sudo mount {0} {1};'
-               'sudo chown -R $USER:$USER {1};').format(device_path, mount_path)
-    self.RemoteCommand(mnt_cmd)
+class BaseOsMixin(object):
+  """The base class for OS Mixin classes.
 
-  def RenderTemplate(self, template_path, remote_path, context):
-    """Renders a local Jinja2 template and copies it to the remote host.
+  This class holds VM methods and attributes relating to the VM's guest OS.
+  For methods and attributes that relate to the VM as a cloud resource,
+  see BaseVirtualMachine and its subclasses.
```

Comment:
Could you document the public attributes of this class (`bootable_time, hostname, remote_access_ports`)?